proposals/044-integration-model.md

```diff
@@ -0,0 +1,203 @@
+Multiple versions of an integration can exist. 
+A _active_ integration is immutable.
+So in order to update an active integration (i.e. an integration which is in state _Active_) a new version needs to be created.
+Changing an _active_ integration in the UI causes a new integration to be created, with the following characteristics:
+* Its cloned from the _active_ integration
+* The state of the clone is set to _Draft_
+* The `parentVersion` is set to the version of the active integration it is cloned from
+* The `version` is set to undefined.
+
+Now the user is able to change anything on this integration until it gets published. When its published the following should happen:
+
+* The `version` is set to a new unique version for this integration
+* The _active_ integration's `targetState` is set to `Undeployed`
+* As soon this state has been reached (callback), this integration's `targetState` is set to `Active`.
+
+That way it is ensured that only a single version of an integration is active. 
+
+## Data model
+
+The following data model only highlights the parts which are specific to the lifecycle model:
+
+```java
+public class Integration {
+    
+    // Unique name of the integration
+    String name;
+    
+    // Current state of the integration
+    IntegrationState currentState;
+    
+    // Target state of the integration    
+    IntegrationState targetState;
+    
+    // The version of the integration which was the origin of this integration
+    // 0 if this is a new integration
+    int parentVersion;
+    
+    // Version of this integratin
+    int version;
+  
+    // Message describing the state further (e.g. error message)
+    String stateMessage;
+    
+    // Whether this integration has converged to its target state
+    boolean isPending() {
+        return currentState != targetState;
+    }
+}
+
+enum IntegrationState {
+  
+    // Initial state of an undeployed integration
+    Draft,
+  
+    // Deployed and running
+    Active, 
+  
+    // Deployed and suspended
+    Inactive,
+  
+    // Undeployed (no associated pods)
+    Undeployed,
+  
+    // Error occured
+    Error;
+}
+```
+
+## Deployment
+
+Deployment goes when an integration with `currentState == Draft` set `targetState = Active`. 
+Several actions will be performed by the controller:
+
+* A GitHub repository will be created if this is the first version of an integration (parentVersion == 0). 
+* If it is an new version of an existing integration (e.g. because a user changed something), a new Git branch is created from the old version's branch. This branch is named after the version.
+* The runtime files are created and commit to the Git repository
+* If this is an update of an existing integration, a currently running or suspended integration (`currentState == Active or Inactive`) is set to `targetState = Undeployed` (which triggers an undeployment on behalf of the controller)
```

Comment:
This kinda seems then like you can't persist changes without publishing them right away, i.e. what if the user wants to make some updates, persist them but not make those changes active right away?